docs/proposal/20201103-failure-domain.md

```diff
@@ -172,138 +173,207 @@ CSI/CPI will be configured with region/zone information so nodes and PVs will ha
+
+  // NetworkName is the network name for this machine's VM.
+  NetworkName string `json:"NetworkName"`
+
+  // Folder is the name or inventory path of the folder in which the
+  // virtual machine is created/located.
+  // +optional
+  Folder string `json:"folder,omitempty"`
+}
 
-    // Network is the network configuration for this machine's VM.
-    Network NetworkSpec `json:"network"`
+type VSphereDeploymentZoneStatus struct {
+  // Ready is true when the VSphereDeploymentZone resource is ready.
+  // If set to false, it will be ignored by VSphereClusters
+  // +optional
+  Ready *bool `json:"ready,omitempty"`
 
-    // Folder is the name or inventory path of the folder in which the
-    // virtual machine is created/located.
-    Folder string `json:"folder"`
+  // Conditions defines current service state of the VSphereMachine.
+  // +optional
+  Conditions clusterv1.Conditions `json:"conditions,omitempty"`
 }
 
-type VSphereClusterSpec struct {
-    // Server is the address of the vSphere endpoint.
-    Server string `json:"server,omitempty"`
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=vspheredeploymentzones,scope=Cluster,categories=cluster-api
+// +kubebuilder:subresource:status
 
-    …..
+// VSphereDeploymentZone is the Schema for the vspheredeploymentzones API
+type VSphereDeploymentZone struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
 
-    // DeploymentZones are names of the VSphereDeploymentZones that
-    // this cluster will use
-    // items in this array should have different failure domains (zones)
-    // if empty, will not use failure domain for control plane nodes
-    DeploymentZones []string `json:"deploymentZones,omitempty"`
+  Spec   VSphereDeploymentZoneSpec   `json:"spec,omitempty"`
+  Status VSphereDeploymentZoneStatus `json:"status,omitempty"`
 }
-
 ```
 
 ### Implementation Details
 
-1. Manually Pre-populate CR for VSphereFailureDomain and VSphereDeploymentZone
-2. Controller change for user story 2/3/4
-3. Controller change for user story 1
-4. New controller for CRD VSphereFailureDomain
-5. Auto discovery, validation of VSphereFailureDomain for user story 1/2/3.
-6. Auto configuration for hostgroup for user story 1/2/3, if provided with enough permission
+the changes required to the controllers are the following
+
+#### VSphereFailureDomain validating webhook
+
+the validation for `VSphereFailureDomain` should:
+
+* verify that the `.Spec.Topology.Datacenter` is not empty
+* if `.Spec.Topology.HostGroup` is not `nil`, verify that `autoconfigure` is not true for both the `hostGroup` and `FailureDomain` struct
+
+#### VSphereDeploymentZone defaulting webhook
+
+the defaultijg for `VSphereDeploymentZone` should:
+
+* check if `controlPlane` is `nil`, if it is default to true
+
+#### the vspheredeploymentzone_controller
+
+This controller would be responsible for:
+
+* Listing `VSphereClusters` and checking if `server` field matches
```

Comment:
would it be worth indexing the server field for easier lookup?

Reply:
ideally we should use match field, but that's up to the implementer

Reply:
yeah match field with the field being indexed would prevent any performance hiccups hopefully